### PR TITLE
Added automated test for custom metrics HPA

### PR DESCRIPTION
**Pull request by marinakog**
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/13824/
https://build.weblogick8s.org:8443/job/wko-oke-nightly-seqential/177 - oke

### CODE REVIEW

**Review by anpanigr — COMMENTED** [review]
Update the description with result form 
    OKE  parallel run and Kind Parallel run

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+import static oracle.weblogic.kubernetes.utils.MonitoringUtils.verifyMonExpAppAccessThroughNginx;
+import static oracle.weblogic.kubernetes.utils.OperatorUtils.installAndVerifyOperator;
+import static oracle.weblogic.kubernetes.utils.PersistentVolumeUtils.createPvAndPvc;
+import static oracle.weblogic.kubernetes.utils.SecretUtils.createSecretWithUsernamePassword;
+import static oracle.weblogic.kubernetes.utils.ThreadSafeLogger.getLogger;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java:85]: Provide a class level javadoc describing the usecase.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+@DisplayName("Test to a create MII domain and test autoscaling using HPA")
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java:86]: The description exactly matches with the class ItHorizontalPodAutoscaler

@DisplayName("Test to a create MII domain and test autoscaling using HPA")

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    monitoringExporterSrcDir = Paths.get(monitoringExporterDir, "srcdir").toString();
+    monitoringExporterAppDir = Paths.get(monitoringExporterDir, "apps").toString();
+    logger.info("Assign a unique namespace for operator");
+    assertNotNull(namespaces.get(0), "Namespace is null");
+    opNamespace = namespaces.get(0);
+
+    logger.info("Assign a unique namespace for WebLogic domain");
+    assertNotNull(namespaces.get(1), "Namespace is null");
+    domainNamespace = namespaces.get(1);
+    logger.info("Get a unique namespace for monitoring");
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java:149]: Get a unique namespace for monitoring --> Get a unique namespace for monitoring Exporter

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+    // check managed servers are up and running for domain1
+    for (int i = 1; i <= replicaCount; i++) {
+      checkPodReadyAndServiceExists(managedServerPrefix + i, domainUid, domainNamespace);
+    }
+  }
+
+
+  /**
+   * Test autoscaling using HPA by increasing the CPU usage.
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java:243]: Need more description with respect to usage of Monitoring Exporter

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java (hunk 1 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java:85]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java (hunk 2 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java (hunk 4 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java (hunk 3 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+  }
+
+
+  /**
+   * Test autoscaling using HPA with custom WLS metrics, collected by Monitoring Exporter
+   * and exposed via Prometheus and Prometheus Adapter.
+   */
+  @DisabledIfEnvironmentVariable(named = "OKD", matches = "true")
+  @Test
+  void testHPAWithMetricsServer() {
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java:249]:
```suggestion
  void testHPAWithCustomMetrics() {
```

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+    List<String> ingressHostList
+        = createIngressForDomainAndVerify(domainUid, domainNamespace, 0, clusterNameMsPortMap,
+        false, ingressClassName, false, 0);
+    // create hpa with custom metrics
+    createHPA();
+    //invoke app 50 times to generate metrics with number of opened sessions > 15
+    verifyMonExpAppAccessThroughNginx(ingressHostList.get(0), 1, nodeportshttp);
+
+    //check hpa scaled up to one more server
+    checkPodReadyAndServiceExists(managedServerPrefix + 3, domainUid, domainNamespace);
```
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java:279]: Does the open sessions go down after some time? can we add check for auto scale down?

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java (hunk 5 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java (hunk 6 above)
Comment (marinakog) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java:292]: added code for scale down check

**Review by vanajamukkara — APPROVED** [review]
LG

**Review by anpanigr — APPROVED** [review]

**Pull request merged**